### ISSUE TITLE
[DSV3] Lack of `norm_topk_prob`

### ISSUE THREAD

**Issue by vwxyzjn** [pytorch/torchtitan#1448]
### Bug description

Thanks for the awesome deepseek v3 implementation. I have been testing it for fine-tuning existing checkpoints. Due to the lack of smaller deepseek v3 checkpoints, I was conducting my tests on 

https://huggingface.co/moonshotai/Moonlight-16B-A3B

However, after the weight conversion, the training grad norm was really high for some reason.

Long story short, it turns out the titan's implementation did not normalize the topk probs of the router scores, whereas the HF implementation [does the normalization ](https://github.com/huggingface/transformers/blob/9b7244f18941bfb9c3ebd5760e36a90722d17a6a/src/transformers/models/deepseek_v3/modeling_deepseek_v3.py#L143)and deepseek's official [inference code does the normalization](https://github.com/deepseek-ai/DeepSeek-V3/blob/f6e34dd26772dd4a216be94a8899276c5dca9e43/inference/model.py#L593) as well.

<img width="1772" height="700" alt="Image" src="https://github.com/user-attachments/assets/61948451-f7a3-4f45-8e79-244fc1c9b740" />

The fix is to add the following

```
class TokenChoiceTopKRouter(nn.Module):
    """This class implements token-choice routing. In token-choice top-K routing, each token is
        routed to top K experts based on the router scores.

    Args:
        gate (nn.Module): Gate module to calculate the scores, typically nn.Linear(dim, num_experts).
        num_experts (int): Number of experts in each moe layer.
        top_k (int): Number of experts each token will be routed to in token-choice routing.
        use_sigmoid (bool): Whether to use sigmoid or softmax for router scores. Default is False.
    """

    def __init__(
        self,
        dim: int,
        num_experts: int,
        top_k: int,
        use_sigmoid: bool = False,
        route_sclaing_factor: float = 1.0,
        norm_topk_prob: bool = False,
    ):
        super().__init__()

        self.dim = dim
        self.num_experts = num_experts
        self.top_k = top_k
        self.use_sigmoid = use_sigmoid
        self.route_sclaing_factor = route_sclaing_factor
        self.norm_topk_prob = norm_topk_prob
        self.gate = nn.Linear(self.dim, self.num_experts, bias=False)

    def forward(
        self, x: torch.Tensor, expert_bias: torch.Tensor | None = None
    ) -> tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
        """
        TODO: We haven't implement the group-based routing (node limit routing),
        and currently EP is not supporting node limit routing yet.

        Args:
            x (torch.Tensor): Input tensor with shape ``(bs*slen, dim)``.

        Returns:
            routed_input (torch.Tensor):
                Tokens grouped together by experts indices with shape ``(bs*slen*top_k,)``.
            token_indices (torch.Tensor):
                Token indices for routed_input with shape ``(bs*slen*top_k,)``.
            num_tokens_per_expert (torch.Tensor):
                Number of tokens assigned to each expert with shape ``(num_experts,)``.
        """
        # scores shape (bs*slen, num_experts)
        scores = self.gate(x)

        # By default, sigmoid or softmax is performed in float32 to avoid loss explosion
        if self.use_sigmoid:
            scores = torch.sigmoid(scores.to(torch.float32))
        else:
            scores = F.softmax(scores.to(torch.float32), dim=1)

        # top scores shape (bs*slen, top_k)
        # Apply expert bias for both selection and final weights (matching HuggingFace behavior)
        if expert_bias is not None:
            scores_with_bias = scores + expert_bias
            _, selected_experts_indices = torch.topk(
                scores_with_bias, k=self.top_k, dim=1
            )
            top_scores = scores_with_bias.gather(dim=1, index=selected_experts_indices)
        else:
            top_scores, selected_experts_indices = torch.topk(
                scores, k=self.top_k, dim=1
            )

        # group tokens together by expert indices from 0 to num_experts and pass that to experts forward
        num_tokens_per_expert = torch.histc(
            selected_experts_indices.view(-1),
            bins=self.num_experts,
            min=0,
            max=self.num_experts,
        )

        # Apply normalization if enabled (before reordering to maintain per-token structure)
        if self.norm_topk_prob:
            # Normalize: divide each token's scores by their sum
            denominator = top_scores.sum(dim=-1, keepdim=True) + 1e-20
            top_scores = top_scores / denominator
            # print(f"[TT Router] Normalization applied: original_sum={denominator[0].item():.6f}, first_scores={top_scores[0]}")
        else:
            # print(f"[TT Router] Normalization NOT applied")
            pass

# [...] existing code
```

Then the grad norm does not explode and the inference is ok.

<img width="2392" height="348" alt="Image" src="https://github.com/user-attachments/assets/55d657c0-8196-470a-9a85-1a1e452a4358" />

<img width="2816" height="1354" alt="Image" src="https://github.com/user-attachments/assets/ffee6489-a31c-4ddd-b29f-29902de2187a" />

### Versions

N/A

**Comment by wwwjn** [comment]
Hi @vwxyzjn thank you so much for sharing this detailed debugging progress and reproduce! And sorry for the confusion, we also received other user's feedback on this and fixed in #1419 . And I'm doing a more detailed numerical verification (layer by layer, even matrix by matrix) by converting HF model weights to torchtitan dsv3 model definition. And I will share the result  soon. Thanks again for bringing this up with such a great reproduce, please double check if #1419 fix the issue!

**Comment by vwxyzjn** [comment]
Oh that's amazing. Thanks @wwwjn that should fix the issue. Feel free to close.

More numerical verification on the inferences / training would be super helpful. Thanks so much @wwwjn!
